File: common/features.py
from contextlib import contextmanager
import json

from common.params import Params

class Features:
  def __init__(self):
    self.p = Params()
    j = json.loads(self.p.get("FeaturesDict", encoding="utf-8"))
    self.dict = j["features"]
    self.packages = j["packages"]

  def clear(self, feature):
    v = int.from_bytes(self.p.get("FeaturesValue"), byteorder="little")
    v = v & (~self.dict[feature])
    self.p.put("FeaturesValue", int.to_bytes(v, byteorder="little", length=8))

  def has(self, feature) -> bool:
    v = int.from_bytes(self.p.get("FeaturesValue"), byteorder="little")
    return (v & self.dict[feature]) != 0

  def reset(self):
    self.p.put("FeaturesValue", b'\x00' * 8)

  def set(self, feature):
    v = int.from_bytes(self.p.get("FeaturesValue"), byteorder="little")
    v = v | self.dict[feature]
    self.p.put("FeaturesValue", int.to_bytes(v, byteorder="little", length=8))

  def set_package(self, package):
    if package not in self.packages:
      return
    self.reset()
    for feature in self.packages[package]:
      self.set(feature)
    self.p.put("FeaturesPackage", package)

File: selfdrive/common/features.cc
#include "selfdrive/common/features.h"

#include "json11.hpp"

Features::Features() {
  std::string err;

  auto dict = json11::Json::parse(Params().get("FeaturesDict"), err);

  static_assert(sizeof(double) >= 8, "double should be big");

  for (auto const& [f, v] : dict["features"].object_items())
    features[f] = (uint64_t) v.number_value();

  for (auto const& [p, l] : dict["packages"].object_items()) {
    packages[p]; // ensure empty entry is inserted
    for (auto const& f : l.array_items())
      packages[p].push_back(f.string_value());
  }
}

void Features::clear(const std::string& feature) {
  uint64_t v = *((uint64_t *) params.get("FeaturesValue").c_str());
  v &= ~(features[feature]);
  params.put("FeaturesValue", (const char *) &v, 8);
}

bool Features::has(const std::string& feature) {
  auto v = *((uint64_t *) params.get("FeaturesValue").c_str());
  return (bool) (v & features[feature]);
}

void Features::reset() {
  uint64_t v = 0;
  params.put("FeaturesValue", (const char *) &v, 8);
}

void Features::set(const std::string& feature) {
  uint64_t v = *((uint64_t *) params.get("FeaturesValue").c_str());
  v |= features[feature];
  params.put("FeaturesValue", (const char *) &v, 8);
}

void Features::set_package(const std::string& package) {
  reset();
  if (packages.find(package) == packages.end())
    return;
  for (auto const& feature : packages[package])
    set(feature);
  params.put("FeaturesPackage", package);
}
